Premium Content / Subscriptions: introduce Paid Newsletter functionality

artpi opened the pull request:
## Ovearching project

The purpose of this branch is to introduce **paid newsletters functionality** to WordPress.com and possibly Jetpack, at a later time. This mostly connects the Memberships module to the Subscriptions module.

You can read more about this work in pbOQVh-2t2-p2

## Purpose of this Pull request

**This branch should have no functional effect in Jetpack.** The feature is gated by `JETPACK_BETA_BLOCKS` flag and JavaScript parts are gated by `paid-newsletters-in-subscriptions` empty extension.

## Changes proposed in this PR


### WP-Admin: Prompt user to create / manage paid plans from the subscribe block:

<img width="1017" alt="Zrzut ekranu 2022-09-27 o 09 31 58" src="https://user-images.githubusercontent.com/3775068/192462696-1214cef9-ff1f-4850-b4d0-089c6c85b710.png">

### WP-Admin: Ability to mark any post as "Paid Subscriber Only" or "Blog Subscribers Only"


![Zrzut ekranu 2022-11-1 o 17 27 34](https://user-images.githubusercontent.com/3775068/199285381-a960c66c-ea40-41a6-86ad-5f403e3aba1d.png)

This functionality will be presented in every blog post on .com and will serve to help paid features discoverability


### Frontend: Open checkout in the subscribe block

After any newsletter-enabled plan is connected to a site, it will open a checkout option before subscribing user:

<img width="1054" alt="Zrzut ekranu 2022-09-27 o 09 35 08" src="https://user-images.githubusercontent.com/3775068/192463200-ef328695-3ddf-42ab-a0b9-8c092564abd4.png">


### Frontend Block and/or unlock access based on JWT token passed from the checkout modal

On the frontend, we **will block access and prompt to subscribe** in following scenarios:

- Post access in post meta `_jetpack_newsletter_access` is set to "All subscribers", user does not have a valid JWT token with `blog_sub: active` property and is not logged in as post editor
- Post access in post meta `_jetpack_newsletter_access` is set to "Paid subscribers", user does not have a valid JWT token with active subscriptions and is not logged in as post editor

![Zrzut ekranu 2022-11-1 o 17 36 52](https://user-images.githubusercontent.com/3775068/199287359-da4adb0d-a3d3-40e8-bf2a-c6e818fa61f6.png)

The complete set of behaviours is described in pbOQVh-2wm-p2

## Testing instructions

For any payment, you want to use card `4242 4242 4242 4242 4242` and any month/year and any cvc combination.

1. Get a Jetpack site. You can use Jurassic Ninja for example: jurassic.ninja/create/?jetpack-beta&branches.jetpack=paid-newsletters
2. Apply this PR
3. Without enabling `JETPACK_BETA_BLOCKS`, this should have no functional changes. 
    - Please put the subscribe block in a post and test the subscription works.
4. ENable  `JETPACK_BETA_BLOCKS`
    -  either via constant in wp-config.php: `define( 'JETPACK_BETA_BLOCKS', true  );`
    - Or if you are on a Jurassic Ninja site, you can set it in UI: /wp-admin/options-general.php?page=companion_settings
5. To sandbox Store:
    - 1. put `define( 'JETPACK__SANDBOX_DOMAIN', implode( '.', [ 'yourwpcomusername', 'dev', 'dfw', 'wordpress', 'com' ] );` in `wp-config.php` in **Jetpack** site. This will point Jetpack to use your sandbox as a sync host
    - 2. put `define( 'USE_STORE_SANDBOX', true );` in your `0-sandbox.php` in your sandbox. This will switch payments to the test cards and you will be able to test payments.
   - 2. Or if you are on a Jurassic Ninja site, you can set it in UI: /wp-admin/options-general.php?page=companion_settings and add your sandbox URL without protocol (see above)  in JETPACK__SANDBOX_DOMAIN
    - 3. Sandbox `public-api.wordpress.com`
6. Prepare your site:
    1. Connect your site to WPCOM using JT or NGROK
    2. Connect your site to Jetpack
    5. Upgrade to a full plan, using the test card details 
    6. Enable setting at the bottom of /wp-admin/admin.php?page=jetpack#/discussion to `Let visitors subscribe to new posts and comments via email` - after that you will be able to see the subscribe block
8. Testing Subscribe block - wp-admin
    1. Insert the subscribe block to a post
    2. Publish it
    3. Go to the post in incognito, submit the form, and see it behave normally - **there should be no functional change**
    4. . You should see "Add a plan" in the subscribe block menubar (in gutenberg). Click it to set up a plan
9. Setting up a plan
    1. It will redirect you to Calypso, where you should be prompted to set up the Stripe connection. If you are sandboxing store, you will see the button to "skip it" - more info in paYJgx-1IE-p2
    2. Once you will have Stripe enabled, you will need to click the link to add a plan again.
    3. Add a plan, make sure the tab with "Newsletter" has enabled the option to subscribe users checking out to the newsletter.
10. Testing subscribe block - frontend
    1. Go on the site frontend in incognito
    2. Put different email than before in the subscribe block, submit
    3. You should see a checkout window with a plan selection after submitting the subscribe form now
    4. Choose a free option, see the checkout window close and the subscribe flow continue without a functional change
    5. Open the subscribe block in incognito AGAIN and insert another email
    6. Choose a paid plan this time, continue with checkout using test card details
    7. After completing everything, the window should close
    8. You should now be a site subscriber. You will be able to see those in https://wordpress.com/people/followers/yoursitedomain
11. Testing post-level controls for a post set to "paid subscribers"
    1. Go to post
    2. Click in the "Newsletter" panel
    3. Select visibility for Paid Subscribers only
    4. Publish a post
    5. Open an incognito window, log in as the user from 10.6 or continue with the checkout flow
    6. Go to a post in incognito, put the email as the user from 10.6
    7. The modal should open, recognize that the user is a paid subscriber and show thank you screen
    8. Click Thank you
    9. The modal should close,
    10. The page should reload and unlock he content
12. Testing post-level controls for a post set to "paid subscribers"
    1. Go to post
    2. Click in the "Newsletter" panel
    3. Select visibility for "All subscribers"
    4. Publish a post
    5. Open an incognito window, log in as the user from 10.4 or continue subscribing (continuing subscribing may be flawed and described in https://github.com/orgs/Automattic/projects/188#card-86514692 )
    6. Go to a post in incognito, put the email as the user from 10.4
    7. The modal should open, recognize that the user is a subscriber and show thank you screen
    8. Click Thank you
    9. The modal should close,
    10. The page should reload and unlock the content

#### Does this pull request change what data or activity we track or use?
No additional information.

n3f committed:
[not verified] Modify the subscribe block by adding 'Add new plan' link. (#26414)

Co-authored-by: brent <brent@n3f.com>

artpi committed:
[not verified] Attach paid subscriptions to subscribe block (#26416)

Open the checkout window if user is using subscribe block and site has paid plans

millerf committed:
[not verified] Added gating for paid newsletter

millerf committed:
[not verified] Fix linting

romarioraffington committed:
[not verified] Add Newsletter Post Visibility Setting (#26412)

* Register post meta to store newsletter visibility

* Add the newsletter settings UI components

*  Use useEntityProp to persist and retrieve meta

millerf committed:
[not verified] Add plans should say "Manage plans" if user has the right plans

artpi committed:
[not verified] Sync meta for post level visibility settings

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/edit.js:
```diff
@@ -204,8 +210,19 @@ export function SubscriptionEdit( props ) {
+
 	return (
 		<>
+			<PluginDocumentSettingPanel title={ __( 'Newsletter settings', 'jetpack' ) }>
```
```suggestion
			<PluginDocumentSettingPanel title={ __( 'Newsletter', 'jetpack' ) }>
```

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/paid-plan.js:
```diff
@@ -0,0 +1,20 @@
+
+export function PaidPlanPanel() {
+	const title = __( 'Paid Newsletter', 'jetpack' );
+	const text = __( 'Create plans for the readers to pay for the content', 'jetpack' );
```
```suggestion
	const text = __( 'Set up a paid plan for readers to access your content.', 'jetpack' );
```

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js:
```diff
@@ -0,0 +1,125 @@
+		label: __( 'Subscribers', 'jetpack' ),
+		info: __( 'Visible to only subscribers', 'jetpack' ),
+	},
+	nobody: {
```
This should be `members_only`, no?

millerf replied:
I think that was copied from GHOST

millerf replied:
I'll change it

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js:
```diff
@@ -0,0 +1,125 @@
+
+export default function NewsletterSettings() {
+	const [ postMeta, setPostMeta ] = useEntityProp( 'postType', 'post', 'meta' );
+	const visibilityKey = postMeta._newsletter_visibility ?? 'public';
```
We need normalize on a consistent naming key that avoids confusion with general post visibility, maybe `jetpack_newsletter_access`.

millerf replied:
Done

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js:
```diff
@@ -0,0 +1,125 @@
+					{ canEdit && (
+						<Dropdown
+							position="bottom left"
+							contentClassName="edit-post-post-visibility__dialog"
```
Is this class needed?

millerf replied:
Yes, for design purposes

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js:
```diff
@@ -0,0 +1,125 @@
+									<fieldset className="editor-post-visibility__fieldset">
+										<VisuallyHidden as="legend">{ __( 'Audience', 'jetpack' ) } </VisuallyHidden>
+										{ Object.keys( visibilityOptions ).map( key => (
+											<div className="editor-post-visibility__choice" key={ key }>
```
Did we look to see if there are existing UI components to handle this?

millerf replied:
I [opened an issue on Gutenberg ](https://github.com/WordPress/gutenberg/issues/44499)so the Visibility component could be componentized and extended here. But no luck for now.

romarioraffington replied:
Yes, Gutenberg doesn't export the PostVisibility component, so we copied it here

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -39,6 +40,29 @@ function register_block() {
 		);
 	}
+
+	register_post_meta(
```
What would be the pros and cons of using `register_post_status()`?

millerf replied:
It is working that way. We could open a discussion about that...

romarioraffington replied:
We agreed that `register_post_status` is the way to go

millerf committed:
[not verified] Change label to "Manage plans" only if there is no newsletter plans

mtias committed:
[not verified] Develop locked content placeholder further.

millerf committed:
[not verified] Fix issue when not on WPCOM

millerf committed:
[not verified] Update composer.lock

artpi committed:
[not verified] Pass the login string to UI

millerf committed:
[not verified] Fix newsletter plans check

artpi committed:
[not verified] Add a check to not kick off purchase if site has no configured newsletter plans

artpi committed:
[not verified] Load Javascript only for sites that have paid plans

millerf committed:
[not verified] Various fixes around 'Add paid plan' button

artpi committed:
[not verified] pass email

millerf committed:
[not verified] Prevent panel to stay open

millerf committed:
[not verified] working version

mtias committed:
[not verified] Adjust copy of paid plan panel

mtias committed:
[not verified] Simplify newsletter panel title

millerf committed:
[not verified] Refactor GetAddPaidPlanButton

millerf committed:
[not verified] Rename paid_members_only to members_only

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js:
```diff
@@ -0,0 +1,125 @@
+				/>
+			</PanelRow>
+
+			<PluginPrePublishPanel title={ __( 'Newsletter settings', 'jetpack' ) }>
```
We should move this to this file: https://github.com/Automattic/jetpack/blob/1565ce1a20bf0eb2ef3a9b69181f11a21caa2175/projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js#L48

millerf replied:
Done

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/paid-plan.js:
```diff
@@ -0,0 +1,18 @@
+	const title = __( 'Paid Newsletter', 'jetpack' );
+	const text = __( 'Set up a paid plan for readers to access your content.', 'jetpack' );
+	return (
+		<>
```
Unnecessary fragment

millerf replied:
Done

artpi committed:
[not verified] fix frontend a bit

millerf committed:
[not verified] remove unecessary fragments

millerf committed:
[not verified] Rename visibility to access

mtias commented on projects/plugins/jetpack/extensions/blocks/subscriptions/edit.js:
```diff
@@ -206,6 +210,9 @@ export function SubscriptionEdit( props ) {
 
 	return (
 		<>
+			<PluginDocumentSettingPanel title={ __( 'Newsletter', 'jetpack' ) }>
```
This should also be moved outside the block functions

millerf committed:
[not verified] Rename visibility to access and move panels

millerf committed:
[not verified] Fix block

artpi committed:
[not verified] change to make overlay work

artpi committed:
[not verified] bug

artpi commented:
Statuses for the post level:
- (empty) | `everybody` -> public access, everybody can access
- `paid_subscribers` -> only for paid subscribers
- `subscribers` -> logged-in site users following in reader

millerf committed:
[not verified] fix gating content

artpi committed:
[not verified] Change the labels of access levels

mtias commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -1034,3 +1035,49 @@ public function register_post_meta() {
+ *
+ * @return string
+ */
+function check_for_paid_subscription( $the_content ) {
```
I think this function should be something like `locked_content`

romarioraffington replied:
Fixed in e5d97e3815a83923b7bbe252b5c86d49a9a5297b. Change to `maybe_get_locked_content`

mtias commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -1034,3 +1035,49 @@ public function register_post_meta() {
+
+	$protection_level = get_post_meta( get_the_ID(), \Automattic\Jetpack\Extensions\Subscriptions\META_NAME_FOR_POST_LEVEL_ACCESS_SETTINGS, true );
+
+	$can_view = true;
```
Should it just return on the if statement?

romarioraffington replied:
Returning early will require more code since we want to return the `do_blocks` code if the user doesn't have access

artpi committed:
[not verified] change labels of access levels

millerf committed:
[not verified] Change naming for level of access

artpi committed:
[not verified] appropriate links for adding a new newsletter plan

millerf committed:
[not verified] Fix default values for access level

romarioraffington committed:
[not verified] Rename function name

n3f committed:
[not verified] Fix unit tests.

n3f committed:
[not verified] Ignore undefined store in tests.

n3f committed:
[not verified] Fix production build.

artpi committed:
[not verified] change the description of a change

artpi committed:
[not verified] rework changelog

artpi commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -1034,3 +1035,54 @@ public function register_post_meta() {
+
+	if ( ! $can_view ) {
+		return do_blocks(
+			'<!-- wp:group {"layout":{"type":"constrained","contentSize":"400px"},"style":{"spacing":{"padding":{"top":"var:preset|spacing|80","right":"var:preset|spacing|80","bottom":"var:preset|spacing|80","left":"var:preset|spacing|80"}}},"backgroundColor":"tertiary"} -->
```
This needs to be cleared up and translated properly

romarioraffington replied:
Can you expound on this or share an example of how do translate this text / block properly?

artpi replied:
1. We need to make it easier to parse somehow - maybe break that placeholder into a separate function? 
2. Proper translation is below - for example `esc_html__( 'Subscribe to get access', 'jetpack' )`;
3. Untranslated elements I can still see are: `"subscribePlaceholder":"Email Address"` and that is probably it.

romarioraffington replied:
Fixed in 
[4b40306](https://github.com/Automattic/jetpack/pull/26417/commits/4b40306f8a0254f7aeae76907bf469d5aacadc0e)

artpi commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -1034,3 +1035,54 @@ public function register_post_meta() {
+	require_once __DIR__ . '/../extensions/blocks/premium-content/_inc/subscription-service/include.php';
+	require_once __DIR__ . '/../modules/memberships/class-jetpack-memberships.php';
+
+	$protection_level = get_post_meta( get_the_ID(), 'jetpack_newsletter_access', true );
```
We need to split off the logic that determines if the user has access to content from here.
Something like `Jetpack_Memberships::current_user_can_view_post`

artpi commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -1034,3 +1035,54 @@ public function register_post_meta() {
+ */
+function maybe_get_locked_content( $the_content ) {
+
+	require_once __DIR__ . '/../extensions/blocks/premium-content/_inc/subscription-service/include.php';
```
We should not require those unless the post meta states that this post is indeed subscriber-only

romarioraffington replied:
Fixed in [303327f](https://github.com/Automattic/jetpack/pull/26417/commits/303327fabd4e2867a7bc9a5ae26e4d8ebbe90114)

artpi commented on projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:
```diff
@@ -453,6 +453,48 @@ public static function is_supported_jetpack_recurring_payments() {
+	 *
+	 * @param string $type - Type of a plan for which site is configured. For now supports empty and newsletter.
+	 */
+	public static function has_configured_plans_jetpack_recurring_payments( $type = '' ) {
```
This function could be combined with the subscequent one in some way - they are duplicating some code

romarioraffington replied:
Not sure how best to combine these two functions as the queries are different

artpi commented on projects/plugins/jetpack/extensions/shared/components/product-management-controls/toolbar-control.js:
```diff
@@ -68,7 +68,7 @@ function NewProduct( { onClose } ) {
 				{ siteSlug && (
 					<ExternalLink
-						href={ `https://wordpress.com/earn/payments-plans/${ siteSlug }#add-new-payment-plan` }
+						href={ `https://wordpress.com/earn/payments-plans/${ siteSlug }#add-newsletter-payment-plan` }
```
Where was it used before this PR?

romarioraffington replied:
This is only used for Payment Blocks when the `CUSTOMIZER_EDITOR` is true for Payment Blocks only. Reverting this change

<img width="1268" alt="Captura de Pantalla 2022-10-25 a las 16 13 49" src="https://user-images.githubusercontent.com/7076981/197883495-291159a0-2e97-44cb-b29f-f1f64f69489f.png">

romarioraffington replied:
Changed in [3e4d486](https://github.com/Automattic/jetpack/pull/26417/commits/3e4d486e95bb857bc54f1a89225c31a56ebab028)

michaeldcain commented on projects/plugins/jetpack/extensions/blocks/subscriptions/settings.js:
```diff
@@ -0,0 +1,104 @@
+	},
+	subscribers: {
+		label: __( 'All subscribers', 'jetpack' ),
+		info: __( 'Visible to everyone to subscribes to your site.', 'jetpack' ),
```
small typo:

```suggestion
		info: __( 'Visible to everyone that subscribes to your site.', 'jetpack' ),
```

romarioraffington replied:
Thanks, resolved in [2701a7c](https://github.com/Automattic/jetpack/pull/26417/commits/2701a7cc3a0d2c902ea4e7656cd8d5136ab037e3)

ebinnion commented:
Matias and I were briefly chatting about the possibility of using post status instead of post meta. I looked into this a bit. It seems simple enough to register a new post type that is public. The concern was around how do we handle these posts if the our functionality was removed, in which case, the post type would no longer be registered. We could change the post types on a deactivation hook or something similar, but that will not be perfect. So, it seems best to stick with post meta.

romarioraffington committed:
[not verified] Correct typo

romarioraffington committed:
[not verified] Move gating logic to Jetpack_Memberships class

romarioraffington committed:
[not verified] Move locked content to function and translate text

romarioraffington committed:
[not verified] Add return type to function

romarioraffington committed:
[not verified] Change hash to add-new-payment-plan

millerf committed:
[not verified] Change meta jetpack_newsletter_access to _jetpack_newsletter_access

romarioraffington commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -15,7 +15,7 @@
 const FEATURE_NAME                             = 'subscriptions';
 const BLOCK_NAME                               = 'jetpack/' . FEATURE_NAME;
-const META_NAME_FOR_POST_LEVEL_ACCESS_SETTINGS = 'jetpack_newsletter_access';
+const META_NAME_FOR_POST_LEVEL_ACCESS_SETTINGS = '_jetpack_newsletter_access';
```
Maybe we should have this constant only stored in one file ( i.e `projects/plugins/jetpack/extensions/blocks/subscriptions/constants.js` ). See [612f6a6](https://github.com/Automattic/jetpack/pull/26417/commits/612f6a632771636ba10a9feceb70b97a9ec8544d)

romarioraffington replied:
We store this meta name [here](https://github.com/Automattic/jetpack/pull/26417/commits/303327fabd4e2867a7bc9a5ae26e4d8ebbe90114#diff-6c4e5f6a52aca3eb00a4cc6c302da733d70d8d26b61d4ba9e4ec2a8141add6e3R40) as well

millerf replied:
Nice catch! Any idea why we are not reusing `META_NAME_FOR_POST_LEVEL_ACCESS_SETTINGS`?!

millerf replied:
> Maybe we should have this constant only stored in one file ( i.e projects/plugins/jetpack/extensions/blocks/subscriptions/constants.js ). See [612f6a6](https://github.com/Automattic/jetpack/pull/26417/commits/612f6a632771636ba10a9feceb70b97a9ec8544d)

I'll take care of that.

romarioraffington replied:
Fixed in [98f00f3](https://github.com/Automattic/jetpack/pull/26417/commits/98f00f36931a02f7ed6da7bbf1e63e8b60be33df)

romarioraffington committed:
[not verified] Remove #add-new-payment-plan from the Manage plans link

romarioraffington committed:
[not verified] Update paid plan panel text based on plan availability

millerf committed:
[not verified] Move constants into their own file and fix missing meta name change

millerf committed:
[not verified] Update composer.lock

Otherwise build won't work

millerf committed:
[not verified] Fix imports

romarioraffington commented on projects/plugins/jetpack/extensions/blocks/subscriptions/edit.js:
```diff
@@ -232,6 +234,11 @@ export function SubscriptionEdit( props ) {
 			</InspectorControls>
+			<BlockControls>
+				<Toolbar>
+					<GetAddPaidPlanButton context={ 'toolbar' } />
```
We may need to add back the `hasNewsletterPlans`  here or pass it to the component.

See https://github.com/Automattic/jetpack/pull/26417/files#diff-e0824d0e12a2d47e8a9baa2bc0533543147d25dfe276b1e0b09fee1b0df1675e

romarioraffington replied:
Fixed in https://github.com/Automattic/jetpack/pull/26417/commits/65e6f12142c9fe4aa51a9680b7d630ab0c88bf05

romarioraffington committed:
[not verified] WIP: Gate content for subscribers

millerf committed:
[not verified] Paid newsletter subscribe access level (#27144)

* Pass post access level to a new var in subscribe.wordpress.com iframe

romarioraffington committed:
[not verified] Fix translation error when builing for production

romarioraffington committed:
[not verified] Pass missing prop to the GetAddPaidPlanButton component

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -106,6 +107,57 @@ public function visitor_can_view_content( $valid_plan_ids ) {
+	 * or the user is logged in and the blog subscription
+	 * is verified by POSTing the Auth endpoint
+	 */
+	public function vistor_is_blog_subscriber() {
```
We need to remove that entire function, we should put that logic here:
https://github.com/Automattic/jetpack/blob/b313bd253afdf919adf7a71a866d0ab85b3a7bda/projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php#L50

We could not be doing any calls here

artpi replied:
I reverted https://github.com/Automattic/jetpack/pull/26417/commits/b909fb0b895d32885a876f938636f5c45f995eae to be continued elsewhere

artpi committed:
[not verified] Newsletters: feature gating (#27142)

* PHP feature gating

* Add feature gating

* fix

* reuse videopress

* fix webpack or other build process stuff

* Add a flag and protect the button on subscribe block

* Prepublish flow

artpi committed:
[not verified] Remove Work-In-Progress for the gating content

This reverts commit b909fb0b895d32885a876f938636f5c45f995eae.

artpi commented on projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js:
```diff
@@ -39,50 +52,61 @@ export default function SubscribePanels() {
-								'This post will be sent to <span>%s readers</span>',
-								subscriberCount,
-								'jetpack'
+				{ showNotices && (
```
Why did we change that?

artpi commented on projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js:
```diff
@@ -39,50 +52,61 @@ export default function SubscribePanels() {
-								'This post has been sent to <span>%s readers</span>',
-								subscriberCount,
-								'jetpack'
+				{ showNotices && (
```
Why did we change that?

n3f committed:
[not verified] Add a link + message for subscription excerpts. (#27198)

FIxes 972-gh-Automattic/multiwoo.

romarioraffington committed:
[not verified] Unlock content based for blog Subscribers (#27212)

* Add access_level to function signature

* Add support for blog_sub parameter

n3f committed:
[not verified] fixing logged in check. (#27282)

romarioraffington committed:
[not verified] Hide comments based on newsletter setting (#27344)

* Hide comments based on newsletter setting
* Add filters for blocking comments

artpi commented on projects/plugins/jetpack/modules/subscriptions/views.php:
```diff
@@ -359,6 +375,7 @@ public static function render_widget_subscription_form( $args, $instance, $subsc
 				method="post"
 				accept-charset="utf-8"
+				data-blog="<?php echo esc_attr( get_current_blog_id() ); ?>"
 				id="<?php echo esc_attr( $form_id ); ?>"
```
Seems like we need `data-post_access_level` here as well

artpi commented on projects/plugins/jetpack/modules/subscriptions/views.php:
```diff
@@ -322,6 +323,21 @@ protected static function get_redirect_fragment( $id = null ) {
+	 */
+	protected static function get_post_access_level() {
+		require_once __DIR__ . '/../../extensions/blocks/subscriptions/constants.php';
+		global $post;
```
Replace with 
https://developer.wordpress.org/reference/functions/get_the_id/

millerf committed:
[not verified] Add/newsletter offline (#27352)

* [not verified] add user_can_receive_post_by_mail

* Updated visitor_can_view_content to user visitor_id && Added user_can_receive_post_by_mail

* Artur's feedback

- Remove $visitor_id parameter
- Remove IS_WPCOM check

* Rework the way access is granted and restructure to centralize the logic

* Feedback https://github.com/Automattic/jetpack/pull/27352/files#r1020302640

* Feedback https://github.com/Automattic/jetpack/pull/27352/files#r1020312430

* Make get_site_id static

* Removed over engineered get_site_id()

* add filter test_jetpack_is_supported_jetpack_recurring_payments for is_supported_jetpack_recurring_payments()

* Make methods not static

* Added tests

* Add post_id to prevent deprecation notice

* Fix tests

* Some clean up

* Some refactor

* [not verified] fix compatibility issue

* [not verified] fix tests

* Fix typo

* [not verified] Refactor

* Remove unecessary $blog_id

* [not verified] Fix tests

* Fix typo

* Reinstate get_site_id()

* [not verified] Added some test for online

* Fix JWT encode routine

* More fix for the tests

* [not verified] Test with admin user

* Small changes

* Added a skipped test for cookies

* Added changelog entry

* Rewrite is_supported_jetpack_recurring_payments()

* Change test name

* Rename and refactor most of the logic for WPCOM online

- Rename  WPCOM_Offline_Subscription_Service to WPCOM_Online_Subscription_Service
- Available across all WPCOM as soon as a user is loggued in
- Added front-end use case
- Added tests for front-end

romarioraffington committed:
[not verified] Add a default param to postMeta (#27508)

useEntityProp can return undefined. To prevent errors add a default state

romarioraffington committed:
[not verified] Unlock content immediately for posts with the subscriber access level (#27433)

*  Unlock content for Jetpack sites subscribers and update const

* No longer need to listen to jetpack_subscribe_continue

* Rely on token to unlock content for free subscribers

* Pass the post_access_level to WPCOM form

This will allow us to customize the options we show in the screen-choose-plan.php template based on the post-access level

romarioraffington committed:
[not verified] Add missing import

millerf committed:
[not verified] Refactor subscription tests (#27564)

* [not verified] Refactor all tests

* Fix linting

* Fix linting and added a test

* [not verified] Fix linting and added a test

* Fix lint

* Remove CamelCase

* Fix tests

* Prevent test failure because of ?? operator

* fix test that were failing

* Ignore CommentedOutCode warning

romarioraffington committed:
[not verified] Prevent submission of empty subscribe field (#27597)

ice9js committed:
[not verified] Fix unit test includes

artpi committed:
[not verified] Skip test failing on php 5.6 and 7.0

ice9js reviewed (commented):
I tried giving this a test run today and unfortunately it fails for me at step `9.iii You should see a checkout window with a plan selection after submitting the subscribe form now`.

When I submit the subscription form the page crashes and I get the following error:
![Screenshot 2022-12-02 at 20 59 54](https://user-images.githubusercontent.com/8056203/205376405-435673f1-abf2-4216-b9c1-819e1544bc6e.png)

I am getting it consistently so it might be something that needs addressing, unless I'm still doing something wrong in which case, please let me know.

millerf committed:
Fix post-merge

millerf committed:
Added dependency for symfony

After merging, I was getting
```
 Problem 1
    - symfony/console is locked to version v6.0.16 and an update of this package was not requested.
    - symfony/console v6.0.16 requires symfony/deprecation-contracts ^2.1|^3 -> could not be found in any version, there may be a typo in the package name.
  Problem 2
    - symfony/service-contracts is locked to version v2.5.2 and an update of this package was not requested.
    - symfony/service-contracts v2.5.2 requires symfony/deprecation-contracts ^2.1|^3 -> could not be found in any version, there may be a typo in the package name.
  Problem 3
    - symfony/console v6.0.16 requires symfony/deprecation-contracts ^2.1|^3 -> could not be found in any version, there may be a typo in the package name.
    - automattic/jetpack-changelogger dev-trunk requires symfony/console ^3.4 || ^5.2 || ^6.0 -> satisfiable by symfony/console[v6.0.16].
    - automattic/jetpack-changelogger is locked to version dev-trunk and an update of this package was not requested.

```

millerf committed:
Remove unused parameter

jeherve committed:
Paid Newsletters: debugging tests (#27905)

* Unskip test

* Run Prettier on panel file

* Remove unecessary comment

* Remove unecessary dependency

* Cast product_id to int for verson <7.2

See https://www.php.net/manual/en/migration72.incompatible.php#migration72.incompatible.object-array-casts

Co-authored-by: Fabien MILLERAND <fab@millerf.com>

millerf committed:
Fix checkout

artpi commented on projects/plugins/jetpack/changelog/add-post-visibility-setting:
```diff
@@ -0,0 +1,5 @@
+Significance: major
+Type: enhancement
+
+Entire post can now be restricted to paid subscribers, or email subscribers. This will enable site owners to reward their paying subscribers with secret Lasagna recipes, or create a paid-only issues of their newsletter.
```
TODO for @artpi : Make this more whimsical

artpi commented on projects/plugins/jetpack/changelog/subscriptions-paid-newsletter:
```diff
@@ -0,0 +1,4 @@
+Significance: major
+Type: enhancement
+
+Paid Newsletter functionality: If the site owner creates one or more "newsletter" paid plans for purchase, subscribers will be asked to pay in the Subscribe block.
```
Again, this could be better

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-online-subscription-service.php:
```diff
@@ -0,0 +1,141 @@
+namespace Automattic\Jetpack\Extensions\Premium_Content\Subscription_Service;
+
+/**
+ * Class WPCOM_Offline_Subscription_Service
```
This is a lie. @millerf I don't follow the rename of `WPCOM_Offline_Subscription_Service` to `WPCOM_Online_Subscription_Service`

millerf replied:
Have a look at the available() function

artpi commented on projects/plugins/jetpack/extensions/blocks/subscriptions/view.js:
```diff
@@ -1 +1,68 @@
+
+domReady( function () {
+	const form = document.querySelector( '.wp-block-jetpack-subscriptions__container form' );
+	let premiumContentJWTToken = '';
```
We also need to bail if the form is not found

artpi committed:
more whimsical description of a change

artpi committed:
bail if form not found

artpi committed:
better description

millerf committed:
Added an offline service

artpi committed:
Remove some styling and shortcode from the subscribe block

artpi committed:
Merge branch 'trunk' into paid-newsletters

artpi committed:
Merge branch 'trunk' into paid-newsletters

millerf committed:
Fix tests

millerf committed:
relaunch tests

ice9js reviewed (dismissed):
I still have issues testing the feature but everything as it should be with the flag turned off. I'd say that's enough to get this merged and resolve the smaller issues individually.

kraftbj committed:
[not verified] Merge remote-tracking branch 'origin/trunk' into paid-newsletters

kraftbj reviewed (dismissed):
Approving based on @ice9js's approval which was dismissed with the extensions.json merge conflict resolution.

jeherve committed:
Update lock file

jeherve committed:
Update lock file back from trunk

romarioraffington committed:
Add missing attributes to form (#28029)

* Add missing attributes to form tag

* Pass wpcom ID and not local blog ID

millerf reviewed (approved):
LGTM (but I might be biais ;))

jeherve reviewed (approved):
Re-approving after another commit.

coder-karen commented:
Dropping a note here that this PR was reverted in https://github.com/Automattic/jetpack/pull/28102 due to fatals during WoA deploy - more information in that PR.
